src/main/scala/spinoco/fs2/mail/internal/internal.scala

```diff
@@ -0,0 +1,43 @@
+package spinoco.fs2.mail
+
+import fs2.{Pipe, Stream}
+import fs2.util.Async
+import fs2.util.syntax._
+
+package object internal {
+
+  /**
+    * Behaves like [[Stream.takeThrough]], but even if down stream ends earlier than the `toggle`
+    * results into false. This will ensure that the [[Stream]] will not end before we get false from the `toggle`.
+    *
+    * This is done by running the source stream in parallel to the consume stream. The messages read from the
+    * source stream are then published into a synchronous queue, that blocks until the down stream reads from it.
+    *
+    * Once downstream finishes this synchronous queue is read until exhaustion marked by [[None]].
+    *
+    */
+  def takeThroughDrain[F[_], A](toggle: A => Boolean)(
```

Comment:
lets name toggle => `f` or `predicate` (similarly like it is for standard takeThrough).